### PR TITLE
#1102 load fewer classes on CDI bootstrap

### PR DESCRIPTION
**Pull request by Emily-Jiang**

### CODE REVIEW

**Comment by tevans78** [comment]
#build

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java [hunk 1]
```diff
@@ -312,11 +316,9 @@ public void scan() throws CDIException {
 
             //and pull out the corresponding classes
             for (String className : rawBeanclassNames) {
-                Class<?> clazz = this.allClasses.get(className);
-                if (clazz != null) {
-                    if (clazz.getClassLoader() == classLoader || !isAccessibleBean(clazz)) {
-                        this.beanClasses.put(className, clazz);
-                    }
+
+                if (this.allClasses.contains(className)) {
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java]: Isn't `this.allClasses` still empty at this point? Does `this.allClasses` ever get fully populated?
p.s. hard to find things in this view!

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java [hunk 2]
```diff
@@ -83,8 +82,9 @@
     private final Set<String> allClassNames = new HashSet<String>();
 
     //sorted maps
-    private final Map<String, Class<?>> allClasses = new TreeMap<String, Class<?>>();
-    private final Map<String, Class<?>> beanClasses = new TreeMap<String, Class<?>>();
+    private final Set<String> allClasses = new HashSet<String>();
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java]: Is there now a difference between this.allClassNames and this.allClasses?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java [hunk 3]
```diff
@@ -150,6 +150,7 @@
         this.id = archiveID;
         this.eeModuleDescptorId = eeModuleDescptorId == null ? archiveID : eeModuleDescptorId;
         this.archive = archive;
+        this.classloader = archive.getClassLoader();
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java]: Notice that there is a parameter on the constructor here which is also called allClasses ... now that they are both Set<String> one could mask the other and so should be renamed.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java [hunk 4]
```diff
@@ -312,11 +316,9 @@ public void scan() throws CDIException {
 
             //and pull out the corresponding classes
             for (String className : rawBeanclassNames) {
-                Class<?> clazz = this.allClasses.get(className);
-                if (clazz != null) {
-                    if (clazz.getClassLoader() == classLoader || !isAccessibleBean(clazz)) {
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java]: I think you still need this bit. Even if you can replace the classloader check using the shadow service, you still need to check if the class already exists in child BDAs (`isAccessibleBean`)

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java [hunk 5]
```diff
@@ -461,11 +462,14 @@ private void scanForEndpoints() throws CDIException {
                     for (ManagedBeanDescriptor<?> managedBeanDescriptor : managedBeanDescriptors) {
                         boolean added = false;
                         for (WebSphereBeanDeploymentArchive child : getWebSphereBeanDeploymentArchives()) {
-                            if ((child.getAllClazzes().containsKey(managedBeanDescriptor.getBeanClass().getName()))
-                                && (child.getAllClazzes().containsValue(managedBeanDescriptor.getBeanClass()))) {
-                                child.addManagedBeanDescriptor(managedBeanDescriptor);
-                                added = true;
-                                break;
+                            if (child.getAllClazzes().contains(managedBeanDescriptor.getBeanClass().getName())) {
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java]: Is it safe to only check by name? I think you need to check that the classloader matches too?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java [hunk 6]
```diff
@@ -476,11 +480,13 @@ private void scanForEndpoints() throws CDIException {
                     for (EjbDescriptor<?> ejbDescriptor : ejbDescriptors) {
                         boolean added = false;
                         for (WebSphereBeanDeploymentArchive child : getWebSphereBeanDeploymentArchives()) {
-                            if ((child.getAllClazzes().containsKey(ejbDescriptor.getBeanClass().getName()))
-                                && (child.getAllClazzes().containsValue(ejbDescriptor.getBeanClass()))) {
-                                child.addEjbDescriptor(ejbDescriptor);
-                                added = true;
-                                break;
+                            if (child.getAllClazzes().contains(ejbDescriptor.getBeanClass().getName())) {
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java]: Is it safe to only check by name? I think you need to check that the classloader matches too?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java [hunk 7]
```diff
@@ -590,26 +596,21 @@ public boolean hasBeans() {
...
 
     @Override
     public Collection<String> getBeanClasses() {
-        return beanClasses.keySet();
+        return beanClasses;
     }
 
     @Override
-    public boolean containsBeanClass(Class<?> clazz) {
+    public boolean containsBeanClass(String clazz) {
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java]: I think a ClassLoader param should be added ... it is the unique combination of class name and classloader that you are looking for

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java [hunk 8]
```diff
@@ -517,7 +517,7 @@ public WebSphereBeanDeploymentArchive getBeanDeploymentArchive(Class<?> beanClas
         // Note this method looks for BDA containing a **bean** of the given class
         // We think this is the correct behavior for the weld CDI11Deployment interfaces
         for (WebSphereBeanDeploymentArchive bda : deploymentDBAs.values()) {
-            if (bda.containsBeanClass(beanClass)) {
+            if (bda.containsBeanClass(beanClass.getName())) {
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java]: Need a classloader check here too

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java [hunk 9]
```diff
@@ -534,10 +534,23 @@ public WebSphereBeanDeploymentArchive getBeanDeploymentArchiveFromClass(Class<?>
...
             for (WebSphereBeanDeploymentArchive bda : orderedBDAs) {
-                if ((bda.getAllClazzes().containsKey(clazz.getName())) && (bda.getAllClazzes().containsValue(clazz))) {
-                    wbda = bda;
-                    classBDAMap.put(clazz, bda);
-                    break;
+                if (bda.getAllClazzes().contains(clazz.getName())) {
+
+                    Class<?> bdaClazz = null;
+                    try {
+                        bdaClazz = CDIUtils.loadClass(bda.getClassLoader(), clazz.getName());
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java]: Although you are asking the BDA's classloader to load the class, that might not be the same classloader that actually does load it. If it isn't then you probably shouldn't add it ... or is the `if (bdaClazz == clazz) {` enough here??

**Comment by Emily-Jiang** [comment]
#build

**Comment by Emily-Jiang** [comment]
#build

**Comment by Emily-Jiang** [comment]
#build

**Comment by Emily-Jiang** [comment]
#build

**Comment by tevans78** [comment]
Passed all FULL FAT tests as well:
https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_q4KyUNstEeeC09PBLf7wmg

**Comment by tevans78** [comment]
superceded by PR #1239

**Pull request closed without merging**